hw/drivers/uart/max3107/src/max3107.c

```diff
@@ -0,0 +1,1090 @@
+                        cfg->clk_source |= CLCSOURCE_PLLBYPASS;
+                        cfg->clk_source &= ~CLCSOURCE_PLLEN;
+                    } else {
+                        cfg->pll_config = pre_div | ((factor_ix - 1) << 6);
+                        cfg->clk_source &= ~CLCSOURCE_PLLBYPASS;
+                        cfg->clk_source |= CLCSOURCE_PLLEN;
+                    }
+                }
+            }
+        }
+    }
+
+    return best_br;
+}
+
+int
+max3107_config_uart(struct max3107_dev *dev, const struct uart_conf_port *conf)
+{
+    int rc;
+
+    if (dev->cfg.crystal_en) {
+        dev->regs.clock.clk_source |= CLCSOURCE_CRYSTALEN;
+    } else {
+        dev->regs.clock.clk_source &= ~CLCSOURCE_CRYSTALEN;
+    }
+    if (dev->cfg.no_pll) {
+        dev->regs.clock.clk_source |= CLCSOURCE_PLLBYPASS;
+    } else {
+        dev->regs.clock.clk_source &= ~CLCSOURCE_PLLBYPASS;
+    }
+    if (dev->cfg.allow_mul_4) {
+        dev->regs.clock.brg_config = BRGCONFIG_4XMODE;
+    } else if (dev->cfg.allow_mul_2) {
+        dev->regs.clock.brg_config = BRGCONFIG_2XMODE;
+    } else {
+        dev->regs.clock.brg_config = 0;
+    }
+    max3107_compute_clock_config(dev->cfg.osc_freq, conf->uc_speed, &dev->regs.clock);
+
+    rc = max3107_write_regs(dev, MAX3107_REG_PLLCONFIG, &dev->regs.clock.pll_config, 5);
+    if (rc) {
+        goto end;
+    }
+
+    dev->regs.modes.mode1 = MODE1_IRQSEL | (conf->uc_flow_ctl ? 0 : MODE1_RTSHIZ);
+    dev->regs.modes.mode2 = 0;
+    dev->regs.modes.lcr = (dev->uart.ud_conf_port.uc_stopbits == 1 ? 0 : LCR_STOPBITS) |
+                          (dev->uart.ud_conf_port.uc_databits - 5) |
+                          ((dev->uart.ud_conf_port.uc_parity != UART_PARITY_NONE) ? LCR_PARITYEN : 0 ) |
+                          ((dev->uart.ud_conf_port.uc_parity == UART_PARITY_EVEN) ? LCR_EVENPARITY : 0);
+    dev->regs.modes.rxtimeout = 2;
+    dev->regs.modes.hdplxdelay = 0;
+    dev->regs.modes.irda = 0;
+
+    rc = max3107_write_regs(dev, MAX3107_REG_MODE1, &dev->regs.modes.mode1, 6);
+    if (rc) {
+        goto end;
+    }
+
+    /* RTS activated at FIFO level 120. */
+    dev->regs.fifo.flow_lvl = 0xFF;
+    dev->regs.fifo.fifo_trg_lvl = (dev->cfg.rx_trigger_level >> 3) << 4 | (dev->cfg.tx_trigger_level >> 3);
+    rc = max3107_write_regs(dev, MAX3107_REG_FLOWLVL, &dev->regs.fifo.flow_lvl, 2);
+    if (rc) {
+        goto end;
+    }
+
+    dev->regs.flow.flow_ctrl = conf->uc_flow_ctl ? (FLOWCTRL_AUTOCTS | FLOWCTRL_AUTORTS) : 0;
+    rc = max3107_write_reg(dev, MAX3107_REG_FLOWCTRL, dev->regs.flow.flow_ctrl);
+    if (rc) {
+        goto end;
+    }
+
+    dev->regs.ints.irq_en = IRQEN_LSRERRIEN | IRQEN_RXTRGIEN;
+    dev->regs.ints.lsr_int_en = LSRINTEN_FRAMEERRIEN | LSRINTEN_PARITYIEN |
+                                LSRINTEN_RBREAKIEN | LSRINTEN_ROVERRIEN | LSRINTEN_RTIMEOUTIEN;
+    rc = max3107_write_regs(dev, MAX3107_REG_IRQEN, &dev->regs.ints.irq_en, 3);
+end:
```

Comment:
this label doesn't do any cleanup etc so maybe avoid all those gotos?

Reply:
if code was to be modified with some kind of tracing single place with return is easier to manage.
My local version (with changes that are for debug purposes only) does just that.

Reply:
alright, fair enough